arch/arm/core/cortex_m/mpu/arm_mpu.c

```diff
@@ -185,6 +185,14 @@ void arm_core_mpu_configure_user_context(struct k_thread *thread)
 {
 	u32_t base = (u32_t)thread->stack_obj;
 	u32_t size = thread->stack_info.size;
+#if !defined(CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT)
+	/* In user-mode the thread stack will include the (optional)
+	 * guard area. For MPUs with arbitrary base address and limit
+	 * it is essential to include this size increase, to avoid
+	 * MPU faults.
+	 */
+	size += thread->stack_info.size - thread->stack_obj;
+#endif
```

Comment:
for user context, we don't need a stack guard.  The guard mechanism is provided by the MPU only allowing access to the user stack.  During system calls, the priv_stack is used.  This needs a guard as we need to have that in place during system calls where the priv_stack is used by the privileged system call.

That said, i have to wonder if we need to keep track of two sizes for the stack.  We dance around the setting of the size to be appropriate for all cases and this has causes some oddities w.r.t. the consistency between MPUs with power of 2 requirements and those without.

Reply:
Well, the situation is more or less, as described. Few more notes.
1. When we create a thread, the thread is in privileged mode, anyways. If stack guards are used, then the first 32 bytes are taken out from size (and `stack_info.start`  points to  `initial stack + 32`). The size is adjusted to remove the guard (32 bytes).
2. On the contrary, `stack_obj `points to the original allocation pointer of the stack object.
3. Now, as @agross-linaro mentions, when we switch to user mode, we do not need guards. However, since the stack region will now start from `stack_obj `, we need to include 32 bytes more into the user region.
4. I think this problem is not visible in v7 MPU, because v7 will round up to the next power of two anyways, but v8 will not.

Reply:
>During system calls, the priv_stack is used. This needs a guard as we need to have that in place during system calls where the priv_stack is used by the privileged system call.

This relates to another possible bug I am currently investigating, @agross-linaro (not related to this patch): When we context-switch, to start a new user thread, the stack guard protects the standard thread stack (because there is not priv stack initialized, yet). However, as soon as the user thread starts, (after _arm_userspace_enter()), the thread may execute system calls. Those calls won't have privileged stack protection (the protection will be enabled in the next context switch). I plan to open a  separate issue for that.